src/newtondynamics_physics/ospnewton.h

```diff
@@ -0,0 +1,124 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <osp/Active/SysPhysics.h>
+#include <osp/Active/ActiveScene.h>
+
+#include <Newton.h>
+
+namespace ospnewton
+{
+
+/**
+ * @brief Associates an entity with a NewtonBody and manages its lifetime
+ */
+struct ACompNwtBody
+{
+
+    ACompNwtBody(NewtonBody const* pBody) noexcept
+     : m_pBody(pBody) { }
+
+    NewtonBody const* body() const noexcept { return m_pBody.get(); }
+
+private:
+    struct Deleter
+    {
+        void operator() (NewtonBody const* pCollision)
+        { NewtonDestroyBody(pCollision); }
+    };
+
+    std::unique_ptr<NewtonBody const, Deleter> m_pBody;
+};
+
+/**
+ * @brief Associates an entity with a NewtonCollision and manages its lifetime
+ */
+struct ACompNwtCollider
+{
+
+    ACompNwtCollider(NewtonCollision const* pCollision) noexcept
+     : m_pCollision(pCollision) { }
+
+    NewtonCollision const* collision() const noexcept
+    { return m_pCollision.get(); }
+
+private:
+    struct Deleter
+    {
+        void operator() (NewtonCollision const* pCollision)
+        { NewtonDestroyCollision(pCollision); }
+    };
+
+    std::unique_ptr<NewtonCollision const, Deleter> m_pCollision;
+};
+
+/**
+ * @brief Represents an instance of a Newton physics world in the scane
+ */
+struct ACtxNwtWorld
+{
+
+    // TODO: not compiling, std::hardware_destructive_interference_size not found?
+    struct alignas(64) PerThread
+    {
+        using SetTfPair_t = std::pair<osp::active::ActiveEnt,
```

Comment:
instead of std::pair, consider just a trivial POD struct, which involves way way less compiler timer than std::pair.

If you still need structured bindings, use https://devblogs.microsoft.com/oldnewthing/20201015-00/?p=104369


e.g.

```cpp
struct Person
{
  std::string name;
  int age;

  template<std::size_t Index>
  auto&& get()       &  { return get_helper<Index>(*this); }

  template<std::size_t Index>
  auto&& get()       && { return get_helper<Index>(*this); }

  template<std::size_t Index>
  auto&& get() const &  { return get_helper<Index>(*this); }

  template<std::size_t Index>
  auto&& get() const && { return get_helper<Index>(*this); }

private:
  template<std::size_t Index, typename T>
  auto&& get_helper(T&& t)
  {
    static_assert(Index < 2,
      "Index out of bounds for Custom::Person");
    if constexpr (Index == 0) return std::forward<T>(t).name;
    if constexpr (Index == 1) return std::forward<T>(t).age;
  }
};
```

Reply:
Doing this in many parts of the code just doesn't seem practical. How about just implementing our own lighter pair?

Reply:
I mean, you still get 70% of the benefits i described by switching to POD structs.

The structured binding support isn't required.

Reply:
Hell, half the reason to support structured bindings is to get better names associated with first and second. Or with tuples, to avoid using std::get<>.

With a POD type, that's not even an issue.